File: unit_test/python_scripts/bert_transformer_test.py
# Copyright 2023 Bytedance Ltd. and/or its affiliates.
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import argparse
import timeit
import random
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F


random.seed(4)
np.random.seed(3)
torch_gen = torch.manual_seed(2)
torch.cuda.manual_seed(1)


def transpose_for_scores(x, n_heads, head_size):
    # (B, S, D) -split-> (B, S, H, W) -trans-> (B, H, S, W)
    new_x_shape = x.size()[:-1] + (n_heads, head_size)
    x = x.view(new_x_shape)
    return x.permute(0, 2, 1, 3)


def seqlen_to_mask(lengths, max_len):
    batch_size = lengths.numel()
    mask = (torch.arange(0, max_len, device=lengths.device)
            .type_as(lengths)
            .repeat(batch_size, 1)
            .lt(lengths.unsqueeze(1)))
    return mask


def set_dtype(ts: torch.Tensor, dtype: str):
    if dtype == "fp32":
        return ts.float()
    elif dtype == "fp16":
        return ts.half()
    raise RuntimeError(f"Unsupported dtype {dtype}")


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("batch_size", type=int)
    parser.add_argument("seqlen", type=int)
    parser.add_argument("head_num", type=int)
    parser.add_argument("head_size", type=int)
    parser.add_argument("--n_layers", default=1, type=int, help="number of transformer layers")
    parser.add_argument("--avg_seqlen", default=0, type=int, help="average seqlen, 0 for equal to seqlen")
    parser.add_argument("--dtype", choices=["fp16", "fp32"], default="fp16")
    parser.add_argument("--export_data", action="store_true", help="whether to export test data. "
                                                                   "if true, only run pytorch to generate data, don't run ByteTransformer")
    parser.add_argument("--export_data_path", default="./", type=str, help="path to export test data")
    parser.add_argument("--lib_path", default="./lib/libths_bytetransformer.so",
                        type=str, help="lib path of torch op ext")
    parser.add_argument("--iters", default=100, type=int, help="perf iterations")
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    batch_size = args.batch_size
    seqlen = args.seqlen
    head_num = args.head_num
    head_size = args.head_size
    avg_seqlen = args.avg_seqlen
    if avg_seqlen <= 0:
        avg_seqlen = seqlen
    n_layers = args.n_layers
    hidden_dim = head_num * head_size
    dtype = args.dtype
    export_data = args.export_data
    export_data_path = args.export_data_path
    lib_path = args.lib_path
    iters = args.iters

    low, high = (2 * avg_seqlen - seqlen, seqlen + 1) if 2 * avg_seqlen > seqlen else (0, 2 * avg_seqlen + 1)
    input_lens = torch.randint(low=low, high=high, size=(batch_size,))
    print("input_lengths:", input_lens)

    seqlen_mask = seqlen_to_mask(input_lens, seqlen)

    # autopep8: off
    qkv_kernel                  = [set_dtype(torch.empty(hidden_dim, hidden_dim * 3).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    qkv_bias                    = [set_dtype(torch.empty(hidden_dim * 3).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    attr_output_kernel          = [set_dtype(torch.empty(hidden_dim, hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    attr_output_bias            = [set_dtype(torch.empty(hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    attr_output_layernorm_gamma = [set_dtype(torch.empty(hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    attr_output_layernorm_beta  = [set_dtype(torch.empty(hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    inter_kernel                = [set_dtype(torch.empty(hidden_dim, hidden_dim * 4).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    inter_bias                  = [set_dtype(torch.empty(hidden_dim * 4).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    output_kernel               = [set_dtype(torch.empty(hidden_dim * 4, hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    output_bias                 = [set_dtype(torch.empty(hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    output_layernorm_gamma      = [set_dtype(torch.empty(hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    output_layernorm_beta       = [set_dtype(torch.empty(hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype) for _ in range(n_layers)]
    from_tensor                 = set_dtype(torch.empty(batch_size, seqlen, hidden_dim).uniform_(-0.4, 0.4).cuda(), dtype)

    attr_mask                   = set_dtype(torch.tile(seqlen_mask, dims=(seqlen,)).reshape(batch_size, seqlen, seqlen).cuda(), dtype)
    # autopep8: on
    is_remove_padding = True
    use_fused_attention = True
    transformer_output = [None for _ in range(n_layers)]

    with torch.no_grad():
        hidden_states = from_tensor
        for layer in range(n_layers):
            input_tensor = hidden_states
            qkv = torch.matmul(hidden_states, qkv_kernel[layer]) + qkv_bias[layer]

            q, k, v = qkv.chunk(3, dim=-1)
            q = transpose_for_scores(q, head_num, head_size)
            k = transpose_for_scores(k, head_num, head_size)
            v = transpose_for_scores(v, head_num, head_size)

            # (B, H, S, W) @ (B, H, W, S) -> (B, H, S, S) -softmax-> (B, H, S, S)
            scores = torch.matmul(q, k.transpose(-2, -1)) / (head_size ** .5)

            scores -= 10000.0 * (1.0 - attr_mask.unsqueeze(1))

            probs = F.softmax(scores, dim=-1)
            # (B, H, S, S) @ (B, H, S, W) -> (B, H, S, W) -trans-> (B, S, H, W)
            h = torch.matmul(probs, v).permute(0, 2, 1, 3).contiguous()
            # -merge-> (B, S, D)
            new_context_layer_shape = h.size()[:-2] + (hidden_dim, )
            hidden_states = h.view(new_context_layer_shape)

            hidden_states = torch.matmul(hidden_states, attr_output_kernel[layer]) + attr_output_bias[layer]

            hidden_states = hidden_states + input_tensor
            hidden_states = F.layer_norm(hidden_states, (hidden_dim, ),
                                         weight=attr_output_layernorm_gamma[layer], bias=attr_output_layernorm_beta[layer])
            residual = hidden_states

            hidden_states = torch.matmul(hidden_states, inter_kernel[layer]) + inter_bias[layer]
            hidden_states = F.gelu(hidden_states)
            hidden_states = torch.matmul(hidden_states, output_kernel[layer]) + output_bias[layer]

            hidden_states = hidden_states + residual
            hidden_states = F.layer_norm(hidden_states, (hidden_dim, ),
                                         weight=output_layernorm_gamma[layer], bias=output_layernorm_beta[layer])

            transformer_output[layer] = hidden_states

    if export_data:
        masked_output = transformer_output[0]
        masked_output = masked_output * set_dtype(seqlen_mask.unsqueeze(-1).cuda(), dtype)

        all_vars = [qkv_kernel[0], qkv_bias[0],
                    attr_output_kernel[0], attr_output_bias[0], attr_output_layernorm_gamma[0], attr_output_layernorm_beta[0],
                    inter_kernel[0], inter_bias[0],
                    output_kernel[0], output_bias[0], output_layernorm_gamma[0], output_layernorm_beta[0],
                    from_tensor, attr_mask, masked_output
                    ]
        file_list = ["qkv_kernel.in", "qkv_bias.in",
                     "attr_output_kernel.in", "attr_output_bias.in", "attr_output_layernorm_gamma.in", "attr_output_layernorm_beta.in",
                     "inter_kernel.in", "inter_bias.in",
                     "output_kernel.in", "output_bias.in", "output_layernorm_gamma.in", "output_layernorm_beta.in",
                     "from_tensor.in", "atten_mask.in", "transformer_out.out"
                     ]
        idx = 0
        for var in all_vars:
            print(str(idx) + " " + file_list[idx] + " " +
                  str(var.shape) + " " + str(var.dtype))
            np.savetxt(file_list[idx], set_dtype(var.flatten(), "fp32").cpu().numpy(), delimiter='\n', fmt='%f')
            idx = idx + 1
    else:
        torch.ops.load_library(lib_path)
        warmup_iters = 5
        for i in range(warmup_iters + iters):
            if i == warmup_iters:
                t0 = timeit.default_timer()

            hidden_states = from_tensor

            for layer in range(n_layers):
                hidden_states = torch.ops.ByteTransformer.BertTransformer(
                    head_num, head_size,
                    qkv_kernel[layer], qkv_bias[layer],
                    attr_output_kernel[layer], attr_output_bias[layer],
                    attr_output_layernorm_gamma[layer], attr_output_layernorm_beta[layer],
                    inter_kernel[layer], inter_bias[layer], output_kernel[layer], output_bias[layer],
                    output_layernorm_gamma[layer], output_layernorm_beta[layer],
                    hidden_states, attr_mask,
                    is_remove_padding, use_fused_attention)

            output = hidden_states

        t1 = timeit.default_timer()
        masked_output = transformer_output[-1]
        masked_output = masked_output * set_dtype(seqlen_mask.unsqueeze(-1).cuda(), dtype)
        print("max diff:", torch.max(torch.abs(masked_output - output)).cpu())
        print("time costs:    {:.2f} ms".format((t1 - t0) * 1000 / iters))

File: unit_test/bert_transformer_test.cc
// Copyright 2023 Bytedance Ltd. and/or its affiliates.
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "bytetransformer/include/bert_transformer.h"
#include "helper.h"
using namespace std;
using namespace bytetransformer;

#ifdef FP16
typedef __half T;
static const OperationType OpType = OperationType::HALF;
#else
typedef float T;
static const OperationType OpType = OperationType::FP32;
#endif

int main(int argc, char *argv[]) {
  if (argc < 5) {
    printf(
        "./bert_transformer_test <batch_size> <seq_len> <head_num> <size_per_head>\n");
    return 0;
  }

  int batch_size = atoi(argv[1]);
  int seq_len = atoi(argv[2]);
  int head_num = atoi(argv[3]);
  int size_per_head = atoi(argv[4]);

  int hidden_dim = head_num * size_per_head;

  bool is_remove_padding = true;
  if (batch_size <= 2) {
    is_remove_padding = false;
  }
  bool use_fused_attention = true;

  if (OpType == OperationType::FP32 || size_per_head != 64 || seq_len > 352)
    use_fused_attention = false;

  bool use_fp32 = true;
  if (OpType == OperationType::FP32)
    use_fp32 = true;

  struct cudaDeviceProp prop;
  cudaGetDeviceProperties(&prop, 0);
  printf("Device %s\n", prop.name);

  cudaStream_t stream;
  check_cuda_error(cudaStreamCreate(&stream));
  cublasHandle_t cublas_handle;
  check_cuda_error(cublasCreate(&cublas_handle));
  check_cuda_error(cublasSetStream(cublas_handle, stream));

  BertTransformerParam<OpType> param;

  FILE *fd = fopen("gemm_config.in", "r");
  int err = 0;
  if (fd == NULL)
    printf("gemm_config.in is not found, using default GEMM algorithms\n");
  else {
    err =
        fscanf(fd, "%d%d%d", &param.cublas_Algo[0], &param.cublas_Algo[1], &param.cublas_Algo[2]);
    err += fscanf(fd, "%d%d", &param.attention_param.cublas_Algo[0],
                  &param.attention_param.cublas_Algo[1]);
    fclose(fd);
    if (err != 5)
      printf("loading GEMM algorithms error, using default GEMM algorithms\n");
  }

  T *qkv_kernel;
  device_malloc(&qkv_kernel, hidden_dim * hidden_dim * 3);
  load_weights(&qkv_kernel, (char *)"qkv_kernel.in", hidden_dim * hidden_dim * 3);
  param.attr_kernel_QKV = qkv_kernel;

  T *qkv_bias;
  device_malloc(&qkv_bias, hidden_dim * 3);
  load_weights(&qkv_bias, (char *)"qkv_bias.in", hidden_dim * 3);
  param.attention_param.attr_bias_QKV = qkv_bias;

  T *attr_output_kernel, *attr_output_bias;
  device_malloc(&attr_output_kernel, hidden_dim * hidden_dim);
  load_weights(&attr_output_kernel, (char *)"attr_output_kernel.in", hidden_dim * hidden_dim);
  device_malloc(&attr_output_bias, hidden_dim);
  load_weights(&attr_output_bias, (char *)"attr_output_bias.in", hidden_dim);
  param.attr_output_kernel = attr_output_kernel;
  param.attr_output_bias = attr_output_bias;

  T *inter_kernel, *inter_bias;
  device_malloc(&inter_kernel, hidden_dim * hidden_dim * 4);
  load_weights(&inter_kernel, (char *)"inter_kernel.in", hidden_dim * hidden_dim * 4);
  device_malloc(&inter_bias, hidden_dim * 4);
  load_weights(&inter_bias, (char *)"inter_bias.in", hidden_dim * 4);
  param.inter_kernel = inter_kernel;
  param.inter_bias = inter_bias;

  T *output_kernel, *output_bias;
  device_malloc(&output_kernel, hidden_dim * 4 * hidden_dim);
  load_weights(&output_kernel, (char *)"output_kernel.in", hidden_dim * 4 * hidden_dim);
  device_malloc(&output_bias, hidden_dim);
  load_weights(&output_bias, (char *)"output_bias.in", hidden_dim);
  param.output_kernel = output_kernel;
  param.output_bias = output_bias;

  if (use_fp32) {
    float *attr_output_layernorm_gamma, *attr_output_layernorm_beta, *output_layernorm_gamma,
        *output_layernorm_beta;
    device_malloc(&attr_output_layernorm_gamma, hidden_dim);
    load_weights(&attr_output_layernorm_gamma, (char *)"attr_output_layernorm_gamma.in",
                 hidden_dim);
    device_malloc(&attr_output_layernorm_beta, hidden_dim);
    load_weights(&attr_output_layernorm_beta, (char *)"attr_output_layernorm_beta.in", hidden_dim);
    device_malloc(&output_layernorm_gamma, hidden_dim);
    load_weights(&output_layernorm_gamma, (char *)"output_layernorm_gamma.in", hidden_dim);
    device_malloc(&output_layernorm_beta, hidden_dim);
    load_weights(&output_layernorm_beta, (char *)"output_layernorm_beta.in", hidden_dim);
    param.attr_output_layernorm_gamma = attr_output_layernorm_gamma;
    param.attr_output_layernorm_beta = attr_output_layernorm_beta;
    param.output_layernorm_gamma = output_layernorm_gamma;
    param.output_layernorm_beta = output_layernorm_beta;
  } else {
    __half *attr_output_layernorm_gamma, *attr_output_layernorm_beta, *output_layernorm_gamma,
        *output_layernorm_beta;
    device_malloc(&attr_output_layernorm_gamma, hidden_dim);
    load_weights(&attr_output_layernorm_gamma, (char *)"attr_output_layernorm_gamma.in",
                 hidden_dim);
    device_malloc(&attr_output_layernorm_beta, hidden_dim);
    load_weights(&attr_output_layernorm_beta, (char *)"attr_output_layernorm_beta.in", hidden_dim);
    device_malloc(&output_layernorm_gamma, hidden_dim);
    load_weights(&output_layernorm_gamma, (char *)"output_layernorm_gamma.in", hidden_dim);
    device_malloc(&output_layernorm_beta, hidden_dim);
    load_weights(&output_layernorm_beta, (char *)"output_layernorm_beta.in", hidden_dim);
    param.attr_output_layernorm_gamma = attr_output_layernorm_gamma;
    param.attr_output_layernorm_beta = attr_output_layernorm_beta;
    param.output_layernorm_gamma = output_layernorm_gamma;
    param.output_layernorm_beta = output_layernorm_beta;
  }

  BertTransformer<OpType> *transformer_layer =
      new BertTransformer<OpType>(batch_size, head_num, size_per_head, seq_len,
                                  use_fused_attention, is_remove_padding, use_fp32);
  transformer_layer->initialize(param);

  void *buf;
  unsigned long long buf_size = transformer_layer->cal_bufsize();
  printf("buf_size %lld bytes, %f in MB\n", buf_size, buf_size / 1024.0 / 1024.0);
  device_malloc((void **)&buf, buf_size);

  T *from_tensor, *atten_mask;
  device_malloc(&from_tensor, batch_size * seq_len * hidden_dim);
  load_weights(&from_tensor, (char *)"from_tensor.in", batch_size * seq_len * hidden_dim);
  device_malloc(&atten_mask, batch_size * seq_len * seq_len);
  load_weights(&atten_mask, (char *)"atten_mask.in", batch_size * seq_len * seq_len);

  T *transformer_out;
  device_malloc(&transformer_out, batch_size * seq_len * hidden_dim);

  struct BertTransformerInferParam<T> infer_param {
    from_tensor, atten_mask, transformer_out, buf, batch_size, seq_len, cublas_handle, stream
  };
  transformer_layer->infer(infer_param);

#ifdef FP16
  printf("\033[0;31m");
  printf("Precision in FP16\n");
  printf("\033[0m");
#else
  printf("\033[0;31m");
  printf("Precision in FP32\n");
  printf("\033[0m");
#endif

  result_check(transformer_out, (char *)"transformer_out.out", batch_size * seq_len * hidden_dim);

  int ite = 50;
  // warm up
  for (int i = 0; i < ite; ++i)
    transformer_layer->infer(infer_param);
  cudaDeviceSynchronize();
  struct timeval ss, ee;
  gettimeofday(&ss, NULL);
  for (int i = 0; i < ite; ++i)
    transformer_layer->infer(infer_param);
  cudaDeviceSynchronize();
  gettimeofday(&ee, NULL);

  printf("transformer costs %.3f ms\n", diffTime(ss, ee) / ite);

  delete transformer_layer;
}
